Committing a Rolled-Back Object (#10899)
mendix/docs: content/en/docs/refguide/modeling/application-logic/microflows-and-nanoflows/activities/object-activities/committing-objects.md

## How Commits Work{#how-commits-work}

In understanding commits, it is important to remember that each persistable object has two states:

1. The state in the database where the values are shared with every other user of the app. This is the state you get when you retrieve an object which is not already in the app's memory.
2. The state in memory where values changed by the app can be seen.

Non-persistable entities behave like the in-memory version of a persistable object.

### Committing Objects

> When you commit an object which is in memory, all changes to the database values are saved. Once it is committed, you cannot roll back to the previous values of the object using the **Rollback object** activity of a microflow.

Reviewer: > all changes to the database values are saved

Do you mean all changes form the in-memory object are saved to the database? Or after the in-memory changes are saved to the database, the database values and in-memory object are in sync? The latter is not necessarily true if some other changes were committed earlier by some other user.

Reviewer: Maybe something like:
> When you commit an object which is in memory, its changes are saved to the database. The object is now unchanged and a roll back will have no effect.

Author: I suggest:

```suggestion
When you commit an object which is in memory, changes to the values are saved in the database. Once it is committed, you cannot roll back to the previous values of the object using the **Rollback object** activity of a microflow.
```

But I want to specifically talk about **Rollback object** as this is different from the rollback which occurs with an error.

However, a Mendix commit is not the same as a database (SQL) COMMIT. When you use a **Commit object(s)** activity, Mendix actually performs an INSERT or UPDATE on the database. For an object of a persistable entity, the database COMMIT is not performed until the microflow and any microflows from which it is called, complete. This means that, although a retrieve from the database by the end-user's app will see the updated version of the object, the updated object will not be seen globally by other end-users until the microflows end.

> Another consequence of this distinction is that, in contrast to an explicit **Rollback object** call, which rolls back to the last Mendix commit, errors in a microflow can initiate a database rollback. If a microflow activity errors and has **Error handling** set to **Rollback** or **Custom with rollback**, the value of the object in the database is rolled back to the value it had at the last savepoint. See [Error Handling in Microflows](/refguide/error-handling-in-microflows/) for more information.

Reviewer: > ... the value of the object in the database is rolled back to the value it had at the last savepoint

Explicitly mention that the in-memory changes are kept?

Author: ```suggestion
Another consequence of this distinction is that, in contrast to an explicit **Rollback object** call, which rolls back to the last Mendix commit, errors in a microflow can initiate a database rollback. If a microflow activity errors and has **Error handling** set to **Rollback** or **Custom with rollback**, the value of the object in the database is rolled back to the value it had at the last savepoint. The object in memory will, however, keep any changes to the values. See [Error Handling in Microflows](/refguide/error-handling-in-microflows/) for more information.
```

## How Commits Work{#how-commits-work}

In understanding commits, it is important to remember that each persistable object has two states:

1. The state in the database where the values are shared with every other user of the app. This is the state you get when you retrieve an object which is not already in the app's memory.
2. The state in memory where values changed by the app can be seen.

Non-persistable entities behave like the in-memory version of a persistable object.

### Committing Objects

When you commit an object which is in memory, all changes to the database values are saved. Once it is committed, you cannot roll back to the previous values of the object using the **Rollback object** activity of a microflow.

> However, a Mendix commit is not the same as a database (SQL) COMMIT. When you use a **Commit object(s)** activity, Mendix actually performs an INSERT or UPDATE on the database. For an object of a persistable entity, the database COMMIT is not performed until the microflow and any microflows from which it is called, complete. This means that, although a retrieve from the database by the end-user's app will see the updated version of the object, the updated object will not be seen globally by other end-users until the microflows end.

Author: I have a contradiction between this statement and [Line 210 of error-handling-in-microflows](https://github.com/mendix/docs/pull/10899/changes#diff-c6263e928b74fedf46fedf9dbdac95a8259bbd6122d7d1cd3e7e0ad7e600da67R210) (which I pulled in from the Mx8 documentation as people said it was useful). In what circumstances does the current end-user continue to see their changes?

If they have uncommitted changes, are these stored in the client and still available if they start a transaction and retrieve the object? Or are they lost if they aren't committed?

This statement implies that the changes continue to be visible to the user even outside the microflow (e.g. if you retrieve into a data container) whereas line 210 of error-handling-in-microflows says the changed data is only visible within the microflow.

I guess this one is more accurate otherwise a **Change Object** with **Commit**: _No_ and **Refresh in Client**: _Yes_ wouldn't work?

Can you confirm?

Reviewer: That line 210 says:
> To ensure that every end-user or process can only see persisted data, data changed in a microflow is only available within that microflow. None of the changes made inside the microflow will be available outside the microflow, not even to the end-user that initiated the microflow. The information will only be available to all end-users of the application once the microflow has successfully completed all the activities.

I do not fully understand what this is meant to say. Changes made by the user in the client are (ephemerally) stored there and applied on in-memory objects when calling a microflow. So the end-user has access to the changes, because they are in the browser.

>  In what circumstances does the current end-user continue to see their changes?

As long as they do not refresh the page and the changes are not committed, these are stored in the client and passed to the runtime when necessary, where they become in-memory objects. When the microflow re-retrieves such an object from the database, this is another copy in memory with the same object id. By itself, this does not interfere with the changes in the client.

However, if the **Refresh in Client** setting is used on the object retrieved from the database, the runtime will tell the client in the response that it should update its state cache with the retrieved object for that id. This will remove the original change.

Returning the retrieved object as the microflow return value keeps the change.

To be honest, I was not 100% sure about the behavior in these 3 scenarios, so I reproduced them in a test project.

Author: Thanks for the investigation. I think there is too much to add to the documentation here, but I have tried to clarify line 210 of error-handling-in-microflows.md in this suggestion: https://github.com/mendix/docs/pull/10899/changes#r3074354106

Reviewer: That change looks ok to me.

    If you want to keep the changes in the version which is in memory, you will have to work around this behavior by creating a new object which contains the changes.

* Deleting an object and then committing it has different outcomes depending on whether the object has already been committed or not:

    * If the object has already been committed, the delete will remove the object from the database, and the subsequent commit will have no effect.
    * If the object in memory is new (that is, it has not been committed before you delete it), the delete will do nothing. However, the subsequent commit will write the object to the database. Therefore, this sequence of actions (a delete followed by a commit) may lead to unexpected results if the object has not been committed before.

#### Committing Non-Persistable Entities

Mendix mimics this behavior for non-persistable entities. This means that:

* Performing a commit on a non-persistable entity means that you cannot use a **Rollback object** activity to go back to the previous values

> * If an error occurs in a microflow, and you have error handling which does a rollback, the values roll back to the state a the previous savepoint.

Author: Rewrite to remove mention of errors.
```suggestion
Mendix mimics this behavior for non-persistable entities. This means that performing a commit on a non-persistable entity means that you cannot use a **Rollback object** activity to go back to the previous values.
```

### Committing Objects

When you commit an object, all changes to the current values are saved. This means that you cannot roll back to the previous values of the object using the **Rollback object** activity of a microflow.

> However, a Mendix commit is not the same as a database (SQL) COMMIT. When you use a **Commit object(s)** activity, Mendix actually performs an INSERT into the database. For an object of a persistable entity, the saved value is not committed to the database until the microflow and any microflows from which it is called, complete. This means that, although the current end user can see the updated version of the object, other end users will not see the updated object until the microflows end.

Reviewer 2: Only if the object is new, does a commit action perform a database `INSERT`. Otherwise, it performs a database `UPDATE`.

Author: Added UPDATE in 89a5f3336953ea3f5b8ba388e6b806e496d294c4

> However, a Mendix commit is not the same as a database (SQL) COMMIT. When you use a **Commit object(s)** activity, Mendix actually performs an INSERT into the database. For an object of a persistable entity, the saved value is not committed to the database until the microflow and any microflows from which it is called, complete. This means that, although the current end user can see the updated version of the object, other end users will not see the updated object until the microflows end.

Reviewer 2: "the saved value is not committed" it still confusing, as it is not clear what type of commit this refers to. After the commit action, the object is in the database, but the transaction has not yet been committed. The effect is that any retrieve from database in that same transaction/microflow will see the (new state of) the object, but other transactions (including, but not limited to, other users) will not see the change. When the microflow ends, the transaction is committed and the change becomes visible globally.

Author: Tried to clarify in 89a5f3336953ea3f5b8ba388e6b806e496d294c4

Reviewer 2: I don't know how much it matters in practice, but strictly speaking the text is still incorrect. Other microflows (or just client retrieves) that belong to that end-user's app will also see the old version until the updating microflow finishes. Until that time, only the updating microflow itself will see the new version.

Another consequence of this is that, in contrast to an explicit **Rollback object** call, errors in a microflow can initiate a rollback. If a microflow activity errors and has **Error handling** set to **Rollback** or **Custom with rollback**, the value of the object is rolled back to the value it had at the start of the microflow. See [Error Handling in Microflows](/refguide/error-handling-in-microflows/) for more information.

#### What Gets Committed

> When you work on an object in memory, Mendix records whether the object has been changed. When you perform a **Commit object(s)** activity, the current value is written to the database and Mendix marks the object as unchanged in memory. This has a couple of consequences that you might not expect:

Reviewer: Earlier in this document, you changed 'current value' to 'changes to current values'. Should you make that distinction here as well?

Author: Updated in ff31fe7c06a52b948f2ee63f92f2661fba363e70

* If you commit an object which is then rolled back due to an error, committing the object again will not write the latest version to the database. You can understand this as the following sequence (see [Error Handling in Microflows](/refguide/error-handling-in-microflows/) for a more detailed discussion of how rollbacks work during error handling):

    1. Your microflow starts and creates a savepoint.
    1. You change your object – it is marked as changed.
    1. You perform a **Commit object(s)** activity which sends the changes to the database – the object is marked as unchanged.

>     1. An error occurs, the microflow ends and data in the database is rolled back to the savepoint.

Reviewer: If I read this correctly, this is an error _after_ the first commit, so unrelated to the commit itself. Maybe make that a bit more explicit as in first passing I read this as the commit itself failed, which results in different behavior.

Author: Updated in ff31fe7c06a52b948f2ee63f92f2661fba363e70

    1. You perform a **Commit object(s)** on the object again, but the changes are not written to the database because:

>         * The object in memory still has your changes

Reviewer: Unless any other changes were made after the first commit, I do not think this is correct. As the commit succeeded, the changes were applied to the object in memory and the object does not have changes. It's just that the earlier changes are now the current values. I.e. rolling back this object would keep the same values.

Reviewer 3: Maybe the next bullet point covers it? Or combine with it?

Author: Combined bullets in ff31fe7c06a52b948f2ee63f92f2661fba363e70

Reviewer: To me, 'the object in memory still has changes' here can mean two things:
- the attributes have the same value as those that you set, but the original values are still there as well. I.e. rolling back this object will undo your changes
- the attributes have the same value as those that you set and the original values are the same. I.e. rolling back this object will keep the same values as the values you set.

        * The object in memory was marked as unchanged after your previous commit
        * The **Commit object(s)** activity does not see the changed marker and so does not recognize that your object in memory has changes which need to be written.

    If you want to keep the changes in the latest version you will have to work around this behavior.

* Deleting an object and then committing it has different outcomes depending on whether the object has already been committed or not:

    * If the object has already been committed, the delete will remove the object from the database, and the subsequent commit will have no effect.
    * If the object is new (that is, it has not been committed before), the delete will do nothing, but the commit will store the object in the database. Therefore, this sequence of actions (a delete followed by a commit) may lead to unexpected results if the object has not been committed before.

#### Committing Non-Persistable Entities

> Mendix mimics this behavior for non-persistable entities. Committing a non-persistable entity means that you cannot use a rollback object activity to go back to the previous values, although the rollback error handling in a microflow rolls back to the original values.

Reviewer: I do not understand this part. Rollback on NPEs and PEs should work the same. After a commit the original values are replaced with the updated values, so a rollback will revert to the updated values. But that is the same for both NPEs and PEs.

Reviewer 2: A transaction rollback (which is what "the rollback error handling in a microflow" refers to) will rollback PEs to their original values from before the microflow, regardless of whether they were committed (as in commit object activity, not as in transaction commit).

Reviewer: I am not sure I understand @basdebakker but as far as I know, in general, nothing of the in-memory state of a mendix object is touched on a transaction commit or rollback. The two are independent.

Author: Should we change this - I'm not sure how to make it clearer. I'm trying to say that PEs and NPEs work the same, but obviously I'm not being clear.

Reviewer 2: What @lvanengelen-mx is saying is just my point (I think). A transaction rollback will rollback any PEs to the state they were in before the transaction started. But it will not affect NPEs. So in that sense they are treated differently.

Reviewer: Bas and I came to the conclusion that we had a misunderstanding on what we meant with PE/NPE: state in memory vs state in the database.

> * If you commit an object which is then rolled back due to an error, committing the object again will not write the latest version to the database. You can understand this as the following sequence (see [Error Handling in Microflows](/refguide/error-handling-in-microflows/) for a more detailed discussion of how rollbacks work during error handling):

Reviewer: > If you commit an object which is then rolled back due to an error

Should we say instead that not _the object_ is rolled back but the transaction is rolled back to the last savepoint?

Author: Good idea:

```suggestion
* If you commit an object and the transaction is then rolled back due to an error, committing the object again will not write the latest version to the database. You can understand this as the following sequence (see [Error Handling in Microflows](/refguide/error-handling-in-microflows/) for a more detailed discussion of how rollbacks work during error handling):
```

    1. Your microflow starts and creates a savepoint.
    1. You change your object in memory – it is marked as changed.
    1. You perform a **Commit object(s)** activity which sends the changes to the database – the object in memory is marked as unchanged.

>     1. An error occurs after the **Commit object(s)** has successfully sent changes to the database – the microflow ends and data in the database is rolled back to the savepoint.

Reviewer: Should we make it more explicit that the error occurs in another action after the commit?

Our implementation of commit could be seen as `Prologue > Database Commit > Epilogue` Any error in `Epilogue` should restore the object to its original state, as that is the contract of the **Commit** action. So this scenario is really about errors in the activity _after_ the commit.

Author: I'll make it more explicit:

```suggestion
    1. An error occurs in an activity after the **Commit object(s)** activity has successfully sent changes to the database – the microflow ends and data in the database is rolled back to the savepoint.
```

    1. You perform a **Commit object(s)** on the object again, but the changes are not written to the database because:

>         * The object in memory still has your changes but it was marked as unchanged after your previous commit

Reviewer: I would avoid 'changes' here, as that implies that it is still in a changed state. Maybe 'committed values' is better?

Author: How about:

```suggestion
        * The object in memory still has the updated values but it was marked as unchanged after your previous commit
```

Author: ```suggestion
        * The object in memory still has the updated values but the attributes were marked as unchanged after your previous commit
```

        * The **Commit object(s)** activity does not see the changed marker and so does not recognize that your object in memory has changes which need to be written.

>     If you want to keep the changes in the version which is in memory, you will have to work around this behavior by creating a new object which contains the changes.

Reviewer: I doubt that creating a new object is correct in general: you probably want it to be the same (i.e. with same object id) object. What would work is update any attribute value (even to its current value). That will mark the in-memory object as changed for the database. Not sure if we want to recommend that to users though...

Author: How about:

```suggestion
    If you want to keep the changes in the version which is in memory, you will have to work around this behavior by modifying the object again.
```

Is this vague enough? I feel like having said that the developer still has the modified version we need to suggest a solution that doesn't involve throwing the changes away and having to do the whole process (which may having involved end-user input) all over again.

Reviewer: I was having doubts about my earlier observation and did an experiment. The change has to be on the attribute that needs to be committed. Changing attribute A won't force a commit on changed attribute B, i.e. the changed state is tracked _per attribute_.

So I like the change that you did but maybe we should use _attributes_ instead of object.

Author: ```suggestion
    If you want to keep the changes in the version which is in memory, you will have to work around this behavior by changing the attributes of the object again.
```

* Deleting an object and then committing it has different outcomes depending on whether the object has already been committed or not:

    * If the object has already been committed, the delete will remove the object from the database, and the subsequent commit will have no effect.
    * If the object in memory is new (that is, it has not been committed before you delete it), the delete will do nothing. However, the subsequent commit will write the object to the database. Therefore, this sequence of actions (a delete followed by a commit) may lead to unexpected results if the object has not been committed before.

#### Committing Non-Persistable Entities

Mendix mimics this behavior for non-persistable entities. This means that:

* Performing a commit on a non-persistable entity means that you cannot use a **Rollback object** activity to go back to the previous values

> * If an error occurs in a microflow, and you have error handling which does a rollback, the values roll back to the state a the previous savepoint.

Reviewer: > the values roll back to the state a the previous savepoint.

The *database* values roll back to the state of the previous savepoint. But as non-persistable entities are not stored in the database, this does not matter for non-persistable entities.

Database rollback and object rollback are unrelated concepts. In-memory objects do not roll back on failure.

Author: So is there no difference for Non-persistable entities between custom with rollback and custom without rollback. They will both behave as custom without rollback? Perhaps I could say that.

```suggestion
* If an error occurs in a microflow, and you have custom error handling, it will always act as a custom without rollback as persistable entities do not have a separate state in the database.
```

Reviewer: > So is there no difference for Non-persistable entities between custom with rollback and custom without rollback. 

I would even say for persistable entities there is no difference between the error handling types. Only because persistable entities are used to interact with the database there are some related effects.

I think your change makes sense by itself, although I have the feeling that if it would be clear in general that database rollback is a different concept to object rollback, this remark would be superfluous.

Author: OK - we won't discuss errors here at all.

```suggestion
```

### Autocommit and Associated Objects {#autocommit-and-associated-objects}

> When an object in memory is committed through a default **Save** button, a commit activity, or web services, it always triggers the commit events. The platform also evaluates all associated objects. To guarantee data consistency, the platform may also autocommit associated objects.

Reviewer: > it always triggers the commit event

Maybe make it explicit that this also holds when the object is not marked as changed (and therefore no actual database save will take place)

Author: I wonder if this, although true, is too much detail. I think "it always" is enough?

Reviewer: Yes, if documented, this behavior should be documented as part of event handling.

> When an object in memory is committed through a default **Save** button, a commit activity, or web services, it always triggers the commit events. The platform also evaluates all associated objects. To guarantee data consistency, the platform may also autocommit associated objects.

Reviewer: > The platform also evaluates all associated objects.

What does this mean?

Author: Good question - I don't actually know what it does but it has some sort of consistency check on all the associated objects. Not sure what else to say here.

Reviewer: If we do not know what it means, do we have to keep it? I tried finding out by looking at the history, but it was already there at the start of the repo it seems.

Author: ```suggestion
When an object in memory is committed through a default **Save** button, a commit activity, or web services, it always triggers the commit events. To guarantee data consistency, the platform may also autocommit associated objects.
```
